android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -1994,7 +1994,25 @@ public static void resetContext(KeyboardType kbType) {
 
   public static int getCurrentKeyboardIndex(Context context) {
     String key = KMKeyboard.currentKeyboard();
-    return KeyboardController.getInstance().getKeyboardIndex(key);
+    // This may be used either _before_ or while the KMKeyboard's host page is still being set up,
+    // either case being before there is an actual value ready for `currentKeyboard`.
+    // (In fact, it may be used to initialize it!)
+    //
+    // So, if there is no current keyboard... we read our cached value for the index of
+    // the user's last selected keyboard.
+    //
+    // For documentation purposes:  "before" can happen for the system keyboard; there's no guarantee
+    // that the host page will finish its async load before the `onStartInput` method is called.
+    if(key == null) {
+      SharedPreferences prefs = context.getSharedPreferences(context.getString(R.string.kma_prefs_name), Context.MODE_PRIVATE);
+      int index = prefs.getInt(KMManager.KMKey_UserKeyboardIndex, 0);
+      if (index < 0) {
+        index = 0;
+      }
```

Comment:
This shifts logic from `pageLoaded` to a shared function which could be called anywhere. I am not very comfortable with that. What about the situation where there are no keyboards loaded when this is called? That's probably not normal for Keyman but it is possible with 3rd party apps that embed KMEA. In this case, index 0 would be an invalid index and cause crashes elsewhere.

Reply:
> What about the situation where there are no keyboards loaded when this is called?

This case is completely unchanged by this PR.  Note how the `if(key == null)` part doesn't actually reference any keyboards, loaded or not -- just the cached index value from "shared preferences".

The part that actually references keyboards was already in the function.

> In this case, index 0 would be an invalid index and cause crashes elsewhere.

That's the very case we're addressing with this PR.  It was already in this function.

Reply:
> What about the situation where there are no keyboards loaded when this is called?

So... issue #6703?  The issue this PR is addressing?  That's the old code in this function.  

Note how the the section - the `if(key == null)` part - doesn't actually reference any keyboards, loaded or not -- just the cached index value from "shared preferences".  It's there _explicitly_ to handle this case.

Following from that, this 'reduces' down to the same concern as your other comment.

Reply:
I don't think you understand what I am saying. If there are no keyboards installed here, we should not be setting an index of `0`. We should be staying with an index of `-1`, which is a valid index, equivalent to `null`.

Reply:
Okay, that definitely makes your underlying question _way_ clearer than before.  Right, I didn't think that much about the 3rd party use case.

Admittedly... I'd been thinking of extracting it to a separate function originally, and then I ran into the nomenclature problem - "what do I name this?"  Turns out the name I chose... already had an implementation.  And it seemed to slot in as a very natural fit, to boot.

Sounds like I should move the `0`-check back to its original location, then.

The line before the block seen above:

```
int index = prefs.getInt(KMManager.KMKey_UserKeyboardIndex, 0);
```

So... should we have this default to `-1` as well, then?  Though... is there really a scenario outside the initial load of the app where there wouldn't be some previously-cached non-negative keyboard index?  Do we actually allow 3rd party apps to do _that_?

Reply:
Yeah, this is tricky: if no default has ever been selected, it probably makes sense to return 0 -- assuming there is a keyboard in that slot. 

A slightly more nuanced, more robust, and probably better solution:

1. If no keyboards installed, return -1.
2. Get the pref value, default 0 (reasonable to default to the first keyboard if never previously checked).
3. If value < 0, make it 0 (so we always try for a valid keyboard).
4. Then, verify that it's a valid keyboard index.
5. If not a valid keyboard index, set to 0 (and update the pref??).

I think that means that this function will always return "good" data which seems like a reasonable design to me 😂: Good data means that if the index >= 0 it is guaranteed at this time to be pointing to an installed keyboard, and -1 always means no keyboards installed.